mlflow/types/schema.py

```diff
@@ -630,28 +660,72 @@ def from_json_dict(cls, **kwargs):
-            raise MlflowException(f"Can't merge map with non-map type: {type(map_type).__name__}")
-        if self == map_type:
+    def _merge(self, other: BaseType) -> Map:
+        if isinstance(other, AnyType) or self == other:
             return deepcopy(self)
+        if not isinstance(other, Map):
+            raise MlflowException(f"Can't merge map with non-map type: {type(other).__name__}")
         if isinstance(self.value_type, DataType):
-            if self.value_type == map_type.value_type:
+            if self.value_type == other.value_type:
                 return Map(value_type=self.value_type)
             raise MlflowException(
                 f"Map types are incompatible for {self} with value_type={self.value_type} and "
-                f"{map_type} with value_type={map_type.value_type}"
+                f"{other} with value_type={other.value_type}"
             )
 
-        if (
-            isinstance(self.value_type, (Array, Object, Map))
-            and self.value_type.__class__ is map_type.value_type.__class__
-        ):
-            return Map(value_type=self.value_type._merge(map_type.value_type))
+        if isinstance(self.value_type, (Array, Object, Map, AnyType)):
+            return Map(value_type=self.value_type._merge(other.value_type))
 
-        raise MlflowException(f"Map type {self!r} and {map_type!r} are incompatible")
+        raise MlflowException(f"Map type {self!r} and {other!r} are incompatible")
+
+
+@experimental
+class AnyType:
+    def __init__(self):
+        """
+        AnyType can store any json-serializable data including None values.
+        For example:
+
+        .. code-block::python
+
+            from mlflow.types.schema import AnyType, Schema, ColSpec
+
+            schema = Schema([ColSpec(type=AnyType(), name="id")])
+
+        .. Note::
+            AnyType should be used when the field is None, the type is not known
+            at the time of data creation, or the field can have multiple types.
+            e.g. for GenAI flavors, the model output could contain `None` values,
+            and `AnyType` can be used to represent them.
+            AnyType has no data validation at all, please be aware of this when
+            using it.
+        """
+
+    def __repr__(self) -> str:
+        return "Any"
+
+    def __eq__(self, other) -> bool:
+        return isinstance(other, AnyType)
+
+    def to_dict(self):
+        return {"type": ANY_TYPE}
+
+    def _merge(self, another_type: BaseType) -> BaseType:
+        if self == another_type:
+            return deepcopy(self)
+        if isinstance(another_type, DataType):
+            return another_type
+        if not isinstance(another_type, BaseType):
+            raise MlflowException(
+                f"Can't merge AnyType with {type(another_type).__name__}, "
+                "it must be a BaseType or DataType"
+            )
+        # Merging AnyType with another type makes the other type optional
```

Comment:
Do we have this side-effect in the current version of the code? (And I guess we shouldn't have it given the definition on an optional field)

Reply:
Yea it shouldn't.
Old behavior: only Nones --> exception; other values + None --> Optional
New behavior: only Nones --> AnyType; other values + None --> Optional